src/extension.ts

```diff
@@ -20,64 +34,172 @@ function getFullTextRange(textEditor: vscode.TextEditor) {
+
+          if (textEditor) {
+            log.appendLine(`Formatting document with '${formatter}'...`)
+
+            const filePath = textEditor.document.fileName
+            const extName = path.extname(filePath)
+            const tmpFilePath = `${path.join(tmpDir, uuid.v4())}${extName}`
+
+            await prepareTmpDir()
+
+            const command = `cd ${rootPath} &&\n  ${formatter} ${filePath} > ${tmpFilePath}`
+            log.appendLine('`' + command + `'`)
+            try {
+              await exec(command)
+            } catch (e) {
+              log.appendLine(e.stderr)
+              vscode.window.showErrorMessage(e.stderr)
+              throw e
+            }
+
+            // TODO: Replace this with `document.getText()`, lest it break Format On Save:
+            //   <https://github.com/microsoft/vscode/issues/90273#issuecomment-584087026>
+            const formattedText = await readFile(tmpFilePath, 'utf8')
+            const textRange = getFullTextRange(textEditor)
+
+            return [vscode.TextEdit.replace(textRange, formattedText)]
+          } else {
+            return []
+          }
+        },
+      },
+    )
+
+  ctx.subscriptions.push(ocamlFormatterDisposable)
+}
 
-  vscode.languages.registerDocumentFormattingEditProvider('reason', {
-    provideDocumentFormattingEdits(document: vscode.TextDocument): vscode.TextEdit[] {
-      const formatterPath = configuration.get<string | undefined>("refmt")
-      const formatter = formatterPath ? path.resolve(rootPath, formatterPath) : "refmt"
-      const textEditor = vscode.window.activeTextEditor
+function registerReasonFormatter(ctx: vscode.ExtensionContext, tmpDir: string) {
+  reasonFormatterDisposable =
+    vscode.languages.registerDocumentFormattingEditProvider(
+      { language: 'reason', scheme: 'file' },
+      {
+        async provideDocumentFormattingEdits(
+          document: vscode.TextDocument,
+        ): Promise<vscode.TextEdit[] | undefined> {
+          const config = vscode.workspace.getConfiguration(
+            'ocaml-reason-format',
+          )
+          if (isDisabled(config)) {
+            log.appendLine(`Extension disabled by user configuration.`)
+            return
+          }
+
+          const rootPath = vscode.workspace.rootPath || ''
+
+          const formatter = config.get<string | undefined>('refmt')
+          const textEditor = vscode.window.activeTextEditor
+
+          if (textEditor) {
+            log.appendLine(`Formatting document with '${formatter}'...`)
+
+            const filePath = textEditor.document.fileName
+            const extName = path.extname(filePath)
+            const tmpFilePath = `${path.join(tmpDir, uuid.v4())}${extName}`
+
+            prepareTmpDir()
+            log.appendLine(`Copying '${filePath}' to '${tmpFilePath}'...`)
+            await copyFile(filePath, tmpFilePath)
+
+            const command = `cd ${rootPath} &&\n  ${formatter} ${tmpFilePath}`
```

Comment:
@ELLIOTTCABLE, can we merge this without solving `getText`? I think it could `formatOnSave` can be resolved with `--in-place` flag or with input redirect like you do with ocamlformat.